README.md

```diff
@@ -174,115 +176,129 @@ lego defaults to communicating with the production Let's Encrypt ACME server. If
-            ]
-        }
-    ]
-}
-```
-
 ## ACME Library Usage
 
 A valid, but bare-bones example use of the acme package:
 
 ```go
+package sample
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"log"
+
+	"github.com/xenolf/lego/acme"
+	"github.com/xenolf/lego/certcrypto"
+	"github.com/xenolf/lego/registration"
+)
+
 // You'll need a user or account type that implements acme.User
 type MyUser struct {
 	Email        string
-	Registration *acme.RegistrationResource
+	Registration *registration.Resource
 	key          crypto.PrivateKey
 }
-func (u MyUser) GetEmail() string {
+
+func (u *MyUser) GetEmail() string {
 	return u.Email
 }
-func (u MyUser) GetRegistration() *acme.RegistrationResource {
+func (u MyUser) GetRegistration() *registration.Resource {
 	return u.Registration
-}
-func (u MyUser) GetPrivateKey() crypto.PrivateKey {
+}*
+func (u *MyUser) GetPrivateKey() crypto.PrivateKey {
 	return u.key
 }
 
-// Create a user. New accounts need an email and private key to start.
-const rsaKeySize = 2048
-privateKey, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
-if err != nil {
-	log.Fatal(err)
-}
-myUser := MyUser{
-	Email: "you@yours.com",
-	key: privateKey,
-}
-
-// A client facilitates communication with the CA server. This CA URL is
-// configured for a local dev instance of Boulder running in Docker in a VM.
-client, err := acme.NewClient("http://192.168.99.100:4000/directory", &myUser, acme.RSA2048)
-if err != nil {
-  log.Fatal(err)
+func main() {
+
+	// Create a user. New accounts need an email and private key to start.
+	const rsaKeySize = 2048
+	privateKey, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	myUser := MyUser{
+		Email: "you@yours.com",
+		key:   privateKey,
+	}
+
+	// A client facilitates communication with the CA server. This CA URL is
+	// configured for a local dev instance of Boulder running in Docker in a VM.
+
+	config := acme.NewDefaultConfig(&myUser).
+		WithCADirURL("http://192.168.99.100:4000/directory").WithKeyType(certcrypto.RSA2048)
+
+	client, err := acme.NewClient(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// We specify an http port of 5002 and an tls port of 5001 on all interfaces
+	// because we aren't running as root and can't bind a listener to port 80 and 443
+	// (used later when we attempt to pass challenges). Keep in mind that we still
+	// need to proxy challenge traffic to port 5002 and 5001.
+	client.Challenge.SetHTTP01Address(":5002")
+	client.Challenge.SetTLSALPN01Address(":5001")
+
+	// New users will need to register
+	reg, err := client.Registration.Register(true)
+	if err != nil {
+		log.Fatal(err)
+	}
+	myUser.Registration = reg
+
+	// The acme library takes care of completing the challenges to obtain the certificate(s).
+	// The domains must resolve to this machine or you have to use the DNS challenge.
+	bundle := true
+	certificates, err := client.Certificate.Obtain([]string{"mydomain.com"}, bundle, nil, false)
```

Comment:
Here, I'd love to just pass in a configuration struct:

```go
client.Certificate.Obtain(certificate.CSR{
    SAN: []string{"mydomain.com"},
    Bundle: true,
    PrivateKey: nil,
    MustStaple: false,
})
```

It's elegant and clearer IMO.

Reply:
As we got `Obtain` and `ObtainCSR` methods, the name of the type of the parameter cannot be `CSR`. I propose `ObtainRequest` as a CQRS oriented naming.

Reply:
Sounds good. Or `ObtainParams`. 🤷‍♂️  :+1: